[RPC Framework] Support passing RemoteModule as an arg

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57876 [RPC Framework] Update rpc.rst
* #57875 [RPC Framework] Add a link to the tutorial in RemoteModule docstring
* **#57695 [RPC Framework] Support passing RemoteModule as an arg**

Add pickling/unpickling support for `RemoteModule`.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28233108](https://our.internmc.facebook.com/intern/diff/D28233108/)

Commit:
[RPC Framework] Support passing RemoteModule as an arg

Add pickling/unpickling support for `RemoteModule`.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28233108](https://our.internmc.facebook.com/intern/diff/D28233108/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -241,6 +252,21 @@ def get_module_rref(self) -> rpc.RRef:
         """
         return self.module_rref
 
+    def __getstate__(self):
```
Add some comments here explaining why we do this.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -33,6 +33,14 @@
     instantiator.instantiate_non_scriptable_remote_module_template()
 )
 
+_REMOTE_MODULE_PICKLED_ATTRIBUTES = (
```
We should also update the docs of RemoteModule mentioning that we can share it across processes using the RPC framework without actually incurring any copy overhead of the actual module and it behaves like an RRef pointing to the remote module.

SciPioneer replied:
Added to the docstring.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -24,6 +25,11 @@ def remote_device(module_rref):
 
 
+# RPC handler for querying __dict__ on the destination worker.
+def remote_module_attributes(remote_module):
```
We should also test that we can actually execute the remote_module on the new host by calling forward and forward_async and verifying it returns the appropriate results.

SciPioneer replied:
Added 2 new tests. However, forward_async cannot be initiated from a new host, because the returned future cannot be pickled.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -385,6 +391,31 @@ def hook(module, grad_input, grad_output):
+            dst_worker_name, modes=[ModuleCreationMode.MODULE_CTOR]
+        ):
+            attrs = rpc.rpc_sync(
+                dst_worker_name, remote_module_attributes, (remote_module,)
```
It looks like we are sharing the remote_module to the process on which the remote_module is already created. We should also test the case where we share it with another process. For example, process 1 creates remote module on process 2, now we share remote module to process 3 and execute it on process 3 (test forward and forward_async) and verify it works correctly.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -385,6 +391,31 @@ def hook(module, grad_input, grad_output):
             ):
                 remote_module.extra_repr()
 
+    @dist_utils.dist_init
```
We should add a test where we validate that if a user adds a new attribute to RemoteModule and it is not in `_REMOTE_MODULE_PICKLED_ATTRIBUTES` we fail. It could be possible that the new attribute might not need to be pickled.

In that case, we could have another list `_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`. In this case, our unit test should validate that each attribute of RemoteModule is either in `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or `_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`. This way it forces developers to add a new attribute to either one of those lists.

SciPioneer replied:
Added _REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING as well as a unit test.

pritamdamania87 reviewed (commented):
Looks good! Have a few comments for additional unit tests.

Commit:
Update on "[RPC Framework] Support passing RemoteModule as an arg"


Add pickling/unpickling support for `RemoteModule`.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28233108](https://our.internmc.facebook.com/intern/diff/D28233108/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Support passing RemoteModule as an arg"


Add pickling/unpickling support for `RemoteModule`.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28233108](https://our.internmc.facebook.com/intern/diff/D28233108/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -241,6 +278,30 @@ def get_module_rref(self) -> rpc.RRef:
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+        # Install generated methods when unpickled.
```
Does this mean the RemoteModule on the remote end will get the forward_async generated method when it is unpickled on the other end? Does generating these methods during unpickling incur any significant overhead?

SciPioneer replied:
> Does this mean the RemoteModule on the remote end will get the forward_async generated method when it is unpickled on the other end?

Yes, only if we pass a remote module as an arg in RPC.

> Does generating these methods during unpickling incur any significant overhead?

The overhead really depends on the cost of running `torch.jit.export` on both `forward` and `forward_async`. I don't have a good answer here. We have already paid this cost and installed these methods on the worker that creates a remote module.

Note that here the scenario is that, if we want to pass this remote module to a new node and let that node initiate forward/forward_async, then we have to do this when unpickling the remote module on the new node. This overhead is not avoidable if the user wants to pass a remote module to an RPC call.

rohan-varma commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -385,6 +404,116 @@ def hook(module, grad_input, grad_output):
+
+        # Unpickled attribtes include both the inherent attributes of RemoteModule
+        # (not inherited from the superclass) and two installed methods.
+        expected_unpickled_attrs = list(_REMOTE_MODULE_PICKLED_ATTRIBUTES)
```
Kind of confused on the name here, its called `expected_unpickled_attrs` but we set it to pickled attributes and the generated methods?

Is it accurate to say that this represents the methods we expect to exist on remote end?

SciPioneer replied:
Kind of confused on the name here, its called expected_unpickled_attrs but we set it to pickled attributes and the generated methods?

I think "methods" can also be viewed as "attributes" in Python, since a method can be set via `setattr`. It's just a bit lengthy to call "expected_unpickled_attrs_and_methods" instead.

> Is it accurate to say that this represents the methods we expect to exist on remote end?

Exactly. So this remote node can access the attributes and the generated methods of the remote node passed from the first node.

rohan-varma commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -385,6 +404,116 @@ def hook(module, grad_input, grad_output):
+            self.assertListEqual(list(attrs.keys()), expected_unpickled_attrs)
+            self.assertEqual(attrs["on"], "worker1")
+            self.assertEqual(attrs["device"], "cpu")
+            self.assertFalse(attrs["is_device_map_set"])
```
should we have a test case where the device map is set and we assert it is set also on the remote end when passed over rpc?

rohan-varma commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
````diff
@@ -385,6 +404,116 @@ def hook(module, grad_input, grad_output):
+
+            # Test the installed methods on worker1's can be initiated by worker2 over RPC layer.
+            # NOTE: In practice a remote module should be directly stored on the worker that runs ``forward``` or ``foward_async``,
+            # not have another worker to initiate forward over the RPC layer.lint
````
nit: ".lint" at end of comment

rohan-varma commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -385,6 +404,116 @@ def hook(module, grad_input, grad_output):
+            # not have another worker to initiate forward over the RPC layer.lint
+            args = (torch.ones(1), 2, "3")
+            ret1 = rpc.rpc_sync(dst_worker2_name, remote_forward, (remote_module, args))
+            self.assertEqual(ret1, tuple(reversed(args)))
```
Can we run a module locally and verify that the results are equivalent?

SciPioneer replied:
I don't think a parity test here is necessary. Some previous tests like `test_forward_async` do not compare it with the result of running locally either. This is because we already know the expected result of local module, and local module is already tested by `nn.Module`.

rohan-varma commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -385,6 +404,116 @@ def hook(module, grad_input, grad_output):
+            self.assertEqual(ret2, tuple(reversed(args)))
+
+    @unittest.skip(
+        "Script RemoteModule cannot be sent over RPC at this time. See #33052"
```
That issue should be resolved and we should be able to send ScriptModules, but not sure about Script Remote Modules. Could you update the comment here and file an issue for this? Also, what exactly is the error that we get?

SciPioneer replied:
Updated with a new issue #57865

rohan-varma commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -24,6 +26,23 @@ def remote_device(module_rref):
+
+# RPC handler for running forward_async on the destination worker.
+def remote_forward_async(remote_module, args):
+    # Since future cannot be pickled and sent over the RPC layer,
```
Can we file an issue for that? Although I guess we can't really support this, besides maybe when the future is set with a value.

SciPioneer replied:
Discussed with Pritam, the users are not supposed to send a future over RPC. Here the purpose is only testing whether the two generated methods are installed on the remote worker.

rohan-varma reviewed (commented):
Looks great overall, thanks! Just a few questions/clarifications about tests

Commit:
Update on "[RPC Framework] Support passing RemoteModule as an arg"


Add pickling/unpickling support for `RemoteModule`.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28233108](https://our.internmc.facebook.com/intern/diff/D28233108/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Thanks for addressing all my questions! Looks good
